packages/transit-vehicle-overlay/src/utils/state.js

```diff
@@ -0,0 +1,160 @@
+import { useState, useRef, useEffect, useCallback } from "react";
+import { checkRefreshInteval } from "./fetch";
+import * as data from "./data";
+
+/**
+ * use this hook when you want your layer to re-paint after otp-ui map zoom events fire
+ * this component will capture and save zoom state from otp-ui's base-map component to state
+ * :return: zoom level (number from leaflet) and the callback for base-map's onViewportChanged
+ */
+export function zoomState(initialZoom = 14) {
+  const [mapZoom, setMapZoom] = useState(initialZoom);
+
+  const onViewportChanged = ({ zoom }) => {
+    // console.info(`zoom level ${zoom}`);
+    setMapZoom(zoom);
+  };
+
+  return [mapZoom, onViewportChanged];
+}
+
+/**
+ * use this hook when you want your layer to re-paint after otp-ui map zoom & recenter events fire
+ * this component will capture and save zoom and center from otp-ui's base-map viewport changes
+ * :return: zoom level, center [x,y] and the callback for base-map's onViewportChanged
+ */
+export function viewState(initialZoom = 14) {
+  const [mapZoom, setMapZoom] = useState(initialZoom);
+  const [mapCenter, setMapCenter] = useState([0.0, 0.0]);
+
+  const onViewportChanged = ({ zoom, center }) => {
+    // console.info(`zoom level ${zoom}`);
+    setMapZoom(zoom);
+    setMapCenter(center);
+  };
+
+  return [mapZoom, mapCenter, onViewportChanged];
+}
+
+/**
+ * trackedVehicleState
+ *
+ * use this hook when you want your layer to track a vehicle
+ * (and potentially show the route geometry of this vehicle)
+ *
+ * @param fetchPatternCallback() that I'll used to fetch a vehicles' pattern geom
+ * @param initial vehicle record
+ * @param initial route pattern
+ * @return [pattern, getTrackedVehicle(), updateTrackedVehicle()]
+ */
+export function trackedVehicleState(
+  fetchPatternCallback = null,
+  initVehicle = null,
+  initPattern = null
+) {
+  const [trackedVehicle, setTrackedVehicle] = useState(initVehicle);
+  const [routePattern, setRoutePattern] = useState(initPattern);
+  const trackedVehicleRef = useRef(trackedVehicle);
+  const routePatternRef = useRef(routePattern);
+
+  // a ref + useEffect give a handle on the current trackedVehicle state in util functions, etc...
+  useEffect(() => {
+    trackedVehicleRef.current = trackedVehicle;
+  }, [trackedVehicle]);
+
+  useEffect(() => {
+    routePatternRef.current = routePattern;
+  }, [routePattern]);
+
+  const getRoutePattern = vehicle => {
+    if (fetchPatternCallback && vehicle) {
+      const pid = routePatternRef.current ? routePatternRef.current.id : null;
```

Comment:
When I see `pid` I think process id. Could you change to pattern Id (or something else if that's not what this is)?